slf4j/src/main/java/com/google/common/flogger/backend/slf4j/Slf4jLoggerBackend.java

```diff
@@ -0,0 +1,179 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.common.flogger.backend.slf4j;
+
+import com.google.common.flogger.backend.LogData;
+import com.google.common.flogger.backend.LoggerBackend;
+import com.google.common.flogger.backend.SimpleMessageFormatter;
+import org.slf4j.Logger;
+
+import java.util.logging.Level;
+
+/**
+ * A logging backend that uses slf4j to output log statements.
+ */
+final class Slf4jLoggerBackend extends LoggerBackend implements
+    SimpleMessageFormatter.SimpleLogHandler {
+
+  private final Logger logger;
+
+  private static final MappedLevel TRACE_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isTraceEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.trace(message, thrown);
+    }
+  });
+
+  private static final MappedLevel DEBUG_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isDebugEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.debug(message, thrown);
+    }
+  });
+  private static final MappedLevel INFO_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isInfoEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.info(message, thrown);
+    }
+  });
+  private static final MappedLevel WARN_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isWarnEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.warn(message, thrown);
+    }
+  });
+  private static final MappedLevel ERROR_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isErrorEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.error(message, thrown);
+    }
+  });
+
+  private static final MappedLevel[] mappedLevels = new MappedLevel[Level.SEVERE.intValue() + 1];
```

Comment:
I'm highly suspicious of making an array with 1001 elements and then adding just 7 values.
See: https://docs.oracle.com/javase/7/docs/api/java/util/logging/Level.html#SEVERE

If this backend can only be configured for one of the in-built levels, then you would have a sorted list such as:

LEVEL_MAP[0] = FINEST.intValue()
LEVEL_MAP[1] = FINER.intValue()
...
LEVEL_MAP[6] = SEVERE.intvalue()

then do a binary search to figure out the "bucket" the given level value is in (a search result of 0 means < FINEST, so no logging, 1 ==> use the FINEST etc...

I'd have thought this could all be done in a switch statement. Something like:

// Result is either >=0 if match an element or (-n - 1) for "insertion point" (but that's just bitwise NOT).
int idx = Arrays.binarySort(LEVEL_MAP, level.intValue());
if (idx < 0) {
  // if insertion point is 'N' then it means the given level value is before the Nth element in the array.
  // The index is ~N, so invert it back but subtract 1 since we want to think of it as "after the previous level".
  idx = ~idx - 1;
}
// idx is now: -1 => no logging, else [0..6] for [FINEST - SEVERE]
switch (idx) {
  case -1:
    // incoming log level lower than FINEST - is there an "ALL" setting for slf4j ?
  case 0:
    // finest logging - check logger is enabled and log.
  ...
  case 6:
    // severe logging or above - log unconditionally? - is there an "OFF" setting for slf4j?
}